Spigot-Server-Patches/0212-Use-TerminalConsoleAppender-for-console-improvements.patch

```diff
@@ -0,0 +1,614 @@
++        // Print message with colors
++        System.out.println(message);
++    }
++
++}
+diff --git a/src/main/java/com/destroystokyo/paper/console/TerminalHandler.java b/src/main/java/com/destroystokyo/paper/console/TerminalHandler.java
+new file mode 100644
+index 00000000..dcd31fbc
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/console/TerminalHandler.java
+@@ -0,0 +1,53 @@
++package com.destroystokyo.paper.console;
++
++import net.minecraft.server.DedicatedServer;
++import net.minecrell.terminalconsole.TerminalConsoleAppender;
++import org.bukkit.craftbukkit.command.ConsoleCommandCompleter;
++import org.jline.reader.LineReader;
++import org.jline.reader.LineReaderBuilder;
++import org.jline.reader.UserInterruptException;
++import org.jline.terminal.Terminal;
++
++public class TerminalHandler {
++
++    private TerminalHandler() {
++    }
++
++    public static boolean handleCommands(DedicatedServer server) {
++        final Terminal terminal = TerminalConsoleAppender.getTerminal();
++        if (terminal == null) {
++            return false;
++        }
++
++        LineReader reader = LineReaderBuilder.builder()
++                .appName("Paper")
++                .terminal(terminal)
++                .completer(new ConsoleCommandCompleter(server))
++                .build();
++        reader.unsetOpt(LineReader.Option.INSERT_TAB);
++
++        TerminalConsoleAppender.setReader(reader);
++
++        try {
++            String line;
++            while (!server.isStopped() && server.isRunning()) {
++                line = reader.readLine("> ");
++                if (line == null) {
++                    break;
++                }
++
++                line = line.trim();
++                if (!line.isEmpty()) {
++                    server.issueCommand(line, server);
++                }
++            }
++        } catch (UserInterruptException e) {
++            server.safeShutdown();
++        } finally {
++            TerminalConsoleAppender.setReader(null);
++        }
++
++        return true;
++    }
++
++}
+diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
+index 2feeb855..db9ace0c 100644
+--- a/src/main/java/net/minecraft/server/DedicatedServer.java
++++ b/src/main/java/net/minecraft/server/DedicatedServer.java
+@@ -73,7 +73,10 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
+                 if (!org.bukkit.craftbukkit.Main.useConsole) {
+                     return;
+                 }
+-                jline.console.ConsoleReader bufferedreader = reader;
++                // Paper start - Use TerminalConsoleAppender implementation
++                if (com.destroystokyo.paper.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
++                BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
```

Comment:
How to change StandardCharsets.UTF_8? In Windows, any input with cyrillic always appears as ╨┐╨░╨▓╨┐╨▓╨...